1-raw-data/mailinglists/getMailData.go

```diff
@@ -0,0 +1,190 @@
+	client, err := storage.NewClient(gcs.ctx)
+	if err != nil {
+		return fmt.Errorf("Failed to create client: %v", err)
+	}
+	gcs.client = client
+	return nil
+}
+
+func (gcs *GCS) createGCSBucket(projectID string) error {
+	// Setup client bucket to work from
+	gcs.bucket = gcs.client.Bucket(gcs.bucketName)
+
+	buckets := gcs.client.Buckets(gcs.ctx, projectID)
+	for {
+		attrs, err := buckets.Next()
+		// Assuming that if end Iterator then not found and need to create
+		if err == iterator.Done {
+			// Create bucket if it doesn't exist - https://cloud.google.com/storage/docs/reference/libraries
+			if err := gcs.bucket.Create(gcs.ctx, projectID, &storage.BucketAttrs{
+				Location: "US",
+			}); err != nil {
+				// TODO - add random number to append to bucket name to resolve
+				return fmt.Errorf("Failed to create bucket:", err)
+			}
+			log.Printf("Bucket %v created.\n", gcs.bucketName)
+			return nil
+		}
+		if err != nil {
+			return fmt.Errorf("Issues setting up Bucket(%q).Objects(): %v. Double check project id.", attrs.Name, err)
+		}
+		if attrs.Name == gcs.bucketName {
+			//getLatestFile() // TODO set this up to check and compare what is in the bucket vs what isn't
+			log.Printf("Bucket %v exists.\n", gcs.bucketName)
+			return nil
+		}
+	}
+}
+
+func (gcs *GCS) storeGCS(fileName string, url string) error {
+	response, err := getHTTPResponse(url)
+	if response.StatusCode == http.StatusOK {
+		obj := gcs.bucket.Object(fileName)
+
+		// w implements io.Writer.
+		w := obj.NewWriter(gcs.ctx)
+
+		// Copy file into GCS
+		_, err := io.Copy(w, response.Body)
+		if err != nil {
+			// TODO simply log failed to copy and store what is missing and keep going - also retry
+			return fmt.Errorf("Failed to copy doc to bucket:", err)
+		}
+		response.Body.Close()
+
+		// Close, just like writing a file.
+		if err := w.Close(); err != nil {
+			return fmt.Errorf("Failed to close:", err)
+		}
+	}
+	return err
+}
+
+func getHTTPResponse(url string) (*http.Response, error) {
+	response, err := http.Get(url)
+	if err != nil {
+		return nil, fmt.Errorf("HTTP response error: ", err)
+	}
+	// Defer response.Body.Close() // TODO figure out if this is usable in a func because its closing when passed
+	return response, nil
+}
+
+func getMailingListData(url string, gcs GCS) error {
+	dom, _ := goquery.NewDocument(url)
+
+	dom.Find("tr").Find("td").Find("a").Each(func(i int, s *goquery.Selection) error {
+		band, ok := s.Attr("href")
+		if ok {
+			check := strings.Split(band, ".")
+			len := len(check) - 1
+			if check[len] == "gz" {
+				if strings.Split(band, ":")[0] != "https" {
+					path := url + band
+					//log.Printf("Relative path to store is: %s\n", path)
+					if err := gcs.storeGCS(band, path); err != nil {
+						return err
+					}
+				}
+			}
+		}
+		return nil
+	})
+	return nil
+}
+
+// TODO create func to create map of what is in bucket and then compare to what is pulled from site so only pull new files
+//func getLatestFile(setup Setup){
```

Comment:
Drop the commented out thing if you can (makes a nice clean commit)

Reply:
Keeping for reference until check is written